cpp/include/cudf/join.hpp

```diff
@@ -0,0 +1,131 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+namespace cudf {
+// joins
+
+/** 
+ * @brief  Performs an inner join on the specified columns of two
+ * dataframes (left, right)
+ *
+ * If join_context->flag_method is set to GDF_SORT then the null_count of the
+ * columns must be set to 0 otherwise @throws GDF Validity is unsupported by sort_join.
+ * 
+ * @example TableA a:{0,1,2}
+ *          TableB b:{1,2,3}, c:{4,5,6} And TableB joins TableA 
+ *          on column 'a' with column 'b'.
+ * The result would be Table-res a:{1, 2}, b:{1, 2}, c:{4, 5}
+ * Result is intersection of column 'a' and column 'b'
+ *
+ * @param[in] left_table The left dataframe
+ * @param[in] right_table The right dataframe
+ * @param[in] join_ind_table The table containing two columns,
+ * first column - indices of left table provided for join
+ * second column - indices of right table provided for join
+ * @param[in] merging_join_ind_table Similar to join_ind_table,
```

Comment:
This still isn't clear. Perhaps it needs an example.

Reply:
This will also contain the left and right indices in two columns, should I also split it in two tables ?

Reply:
And for example of what it has,

`
import cudf
import pandas as pd

df1 = cudf.DataFrame({'a':[0,1,2.0,3,4,5, 9], 'c':[0,2,3,5,7,8,9], 'd':range(7)})
df2 = cudf.DataFrame({'a':[12,5.0,3,9,5], 'c':range(5), 'e':range(5)})

res = df1.merge(df2, how='left', left_on=['a', 'd'], right_on=['a', 'e'])
print(res)
`

In this case, this table will have two columns [0], [0], which indicate the series with name 'a' in both dataframe and are being referenced to join at same index in 'on', and they will end up being single column in result.

Reply:
Okay, this definitely should not be a `table`.

It should be a `std::vector<std::pair>` that contains the column indices from `left_on` and `right_on` that are the same and will only materialize a single result column.

Reply:
ok

Reply:
That's much less intuitive and it decouples information that should be linked. We want to group together the relevant information of a column index from the `left_on` and `right_on` tables that are the same. This is what a `vector` of index `pair`s  does this. A `pair` of `vector`s requires accessing the same element from from disjoint vectors, and the information is no longer coupled.